File: cpp/src/FileMonitor.cpp
#include "../include/FileMonitor.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

FileMonitor::FileMonitor()
{
}

FileMonitor::FileMonitor(string file) : _file(file)
{
}

void FileMonitor::setLocked(bool lock)
{
    _lock = lock;
}

bool FileMonitor::isLocked()
{
    return _lock;
}

void FileMonitor::run()
{
    cout << "Call to FileMonitor::run()" << endl;
    struct stat fileStruct;

    if(stat(_file.c_str(), &fileStruct) == -1) {
        perror("stat");
    }

    long timestamp = fileStruct.st_mtime;

    if(_timestamp != timestamp) {
        _timestamp = timestamp;

        if(!isLocked())
            onChange();
    }
}

File: ats/src/main/java/fr/automated/trading/systems/tradingrobot/monitor/filemonitoring/FileMonitor.java
package fr.automated.trading.systems.tradingrobot.monitor.filemonitoring;

import fr.automated.trading.systems.tradingrobot.monitor.IMonitor;

import java.io.File;
import java.util.TimerTask;

public abstract class FileMonitor extends TimerTask implements IMonitor {
	
	private long timeStamp = 0;
	private boolean lock = false;
	private final File file;
	
	public FileMonitor(File file) {
		this.file = file;
	}
	
	public void run() {

		long timeStamp = file.lastModified();

		if(this.timeStamp != timeStamp) {
			this.timeStamp = timeStamp;

			if(!isLocked())
				onChange();
	    }
	}

	public abstract void onChange();
	
	protected void setLocked(boolean lock) {
		this.lock = lock;
	}

	protected boolean isLocked() {
		return lock;
	}
}